src/credentials/providers/sharedCredentialsProvider.ts

```diff
@@ -0,0 +1,192 @@
+/*!
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as AWS from 'aws-sdk'
+import { Profile } from '../../shared/credentials/credentialsFile'
+import { getLogger } from '../../shared/logger'
+import { getStringHash } from '../../shared/utilities/textUtilities'
+import { getMfaTokenFromUser } from '../credentialsCreator'
+import { CredentialsProvider } from './credentialsProvider'
+import { makeCredentialsProviderId } from './credentialsProviderId'
+
+const SHARED_CREDENTIAL_PROPERTIES = {
+    AWS_ACCESS_KEY_ID: 'aws_access_key_id',
+    AWS_SECRET_ACCESS_KEY: 'aws_secret_access_key',
+    AWS_SESSION_TOKEN: 'aws_session_token',
+    CREDENTIAL_PROCESS: 'credential_process',
+    REGION: 'region',
+    ROLE_ARN: 'role_arn',
+    SOURCE_PROFILE: 'source_profile'
+}
+
+/**
+ * Represents one profile from the AWS Shared Credentials files, and produces Credentials from this profile.
+ */
+export class SharedCredentialsProvider implements CredentialsProvider {
+    private static readonly CREDENTIALS_TYPE = 'profile'
+
+    private readonly profile: Profile
+
+    public constructor(
+        private readonly profileName: string,
+        private readonly allSharedCredentialProfiles: Map<string, Profile>
+    ) {
+        const profile = this.allSharedCredentialProfiles.get(profileName)
+
+        if (!profile) {
+            throw new Error(`Profile not found: ${profileName}`)
+        }
+
+        this.profile = profile
+    }
+
+    public getCredentialsProviderId(): string {
+        return makeCredentialsProviderId({
+            credentialType: SharedCredentialsProvider.getCredentialsType(),
+            credentialTypeId: this.profileName
+        })
+    }
+
+    public getHashCode(): number {
+        return getStringHash(JSON.stringify(this.profile))
+    }
+
+    public getDefaultRegion(): string | undefined {
+        return this.profile[SHARED_CREDENTIAL_PROPERTIES.REGION]
+    }
+
+    /**
+     * Returns undefined if the Profile is valid, else a string indicating what is invalid
+     */
+    public validate(): string | undefined {
```

Comment:
What happens if the profile has neither ROLE_ARN nor CREDENTIAL_PROCESS nor AWS_SESSION_TOKEN nor AWS_ACCESS_KEY_ID?

Reply:
It isn't necessarily invalid, but it is unsupported. Currently, the toolkit would allow the user to see and select it, but it would log "did not contain any supported properties" and throw an Error after the user selects it, when getting credentials.

What should we do here? I'm wondering if we should validate "unsupported" entries like this out in the validate call.

Reply:
The JetBrains toolkit considers this an invalid profile (unsupported). I'll update things to treat it similarly.